### PR TITLE
feat: validate caveat payloads

### PR DESCRIPTION
**Pull request by alanshaw**
This allows caveats (nb) to be used in server handler functions.

It adds a minimal validator framework that parses caveats to a Go type.

You can now define a capability with a Go type for caveats (and an IPLD schema). This allows incoming capability invocation payloads to be parsed to check for validity.

### CODE REVIEW

**Comment by alanshaw** [comment]
@hannahhoward I merged the `Rebind` changes here - please review the rest when you have a sec 🙏

File: ucan/lib.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ import (
 	"time"
 
 	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/storacha-network/go-ucanto/core/ipld"
```
Comment (hannahhoward) [inline ucan/lib.go:8]: I wonder if we should just redefine the builder interface here.

Reason:
1. This links the ucan package back into ucanto core, which makes it less easily extractable.
2. There's minimal consequence to doing so -- two interfaces with the same methods are essentially interchangable.

File: core/result/result.go [hunk 2]
```diff
@@ -10,6 +10,7 @@ import (
 )
 
 // Result is a golang compatible generic result type
+// https://github.com/ucan-wg/receipt/#6-result
```
Comment (hannahhoward) [inline core/result/result.go:13]: I took this out just cause it wasn't actually matching  the UCAN type any more

File: ucan/lib.go (hunk 1 above)
Comment (hannahhoward) [reply ucan/lib.go:8]: re: extractability -- i am in active conversations with folks who want this extracted to become the default go-ucan, especially once it's 1.0

**Review by hannahhoward — APPROVED** [review]
I think I'm LGTM on the rest of this, but I  we should take steps to insure the ucan package does not link to anything else (either by copying the ipld.builder interface, or just putting it in the ucan package)

File: ucan/lib.go (hunk 1 above)
Comment (alanshaw) [reply ucan/lib.go]: > I wonder if we should just redefine the builder interface here.

I did this 👍

**Pull request merged**